### PR TITLE
Support for replacement of a downstairs.

### PR DESCRIPTION
**Pull request by leftwo**
This adds support for the replacement of a downstairs while the upstairs is active and taking IOs.  We make use of the LiveRepair code to do the actual data transfer to the new downstairs.

Two new states were added for a downstairs: Replacing and Replaced. When requested, the upstairs will move a downstairs into Replacing. This starts the process, and all tasks will quit their current work and eventually force a disconnect from the existing downstairs.  This then triggers a connect to the replacement downstairs which then is sent through the Live Repair code path to get that downstairs matching.

For crutest:
Added a replace command to the CLI,
Added two new tests, replace and replay.

The replay test uses dsc to stop downstairs, then send some IO then restart that downstairs.

The replace test uses the new guest.downstairs_replace method to replace a downstairs while sending IO and verifies the IO all comes back as expected.

Both replace and replay can run for a specific number of rounds, or until they receive a signal.  These two tests should be able to replace some of the shameful bash tests we have currently.

Updated and added additional downstairs replacement tests to the integration_tests suite.  Removed some debug cruft that had slipped in at some point.

Added some replace downstairs tests to the upstairs tests.

### CODE REVIEW

**Comment by leftwo** [comment]
The crutest replace test ran all night on two machines, no hangs or problems found.  Marking this ready for review.

**Comment by leftwo** [comment]
Here is some sample output from the dtrace count script showing downstairs moving through the 
`Active` -> `Replacing` -> `Replaced` -> `LiveRepairReady` -> `LiveRepair` -> `Active` cycle.

Because the dtrace script checks ever 5 seconds, you don't always see every transition.
```
           active            active            active   386  386  385  385  385  384    0    0    0  385  385  384
           active            active         replacing   386  386  385  385  385  384    0    0    0  385  385  385
           active            active         replacing   386  386  385  385  385  384    0    0    0  385  385  385
       DS STATE 0        DS STATE 1        DS STATE 2  CON0 CON1 CON2 LRC0 LRC1 LRC2 LRA0 LRA1 LRA2 REP0 REP1 REP2
           active            active       live_repair   386  386  386  385  385  384    0    0    0  385  385  385
           active            active       live_repair   386  386  386  385  385  384    0    0    0  385  385  385
           active            active       live_repair   386  386  386  385  385  384    0    0    0  385  385  385
           active            active       live_repair   386  386  386  385  385  384    0    0    0  385  385  385
           active            active            active   386  386  386  385  385  385    0    0    0  385  385  385
           active            active            active   386  386  386  385  385  385    0    0    0  385  385  385
         replaced            active            active   386  386  386  385  385  385    0    0    0  386  385  385
      live_repair            active            active   387  386  386  385  385  385    0    0    0  386  385  385
      live_repair            active            active   387  386  386  385  385  385    0    0    0  386  385  385
      live_repair            active            active   387  386  386  385  385  385    0    0    0  386  385  385
      live_repair            active            active   387  386  386  385  385  385    0    0    0  386  385  385
           active            active            active   387  386  386  386  385  385    0    0    0  386  385  385
           active            active            active   387  386  386  386  385  385    0    0    0  386  385  385
           active            active            active   387  386  386  386  385  385    0    0    0  386  385  385
           active            active            active   387  386  386  386  385  385    0    0    0  386  385  385
           active         replacing            active   387  386  386  386  385  385    0    0    0  386  386  385
           active         replacing            active   387  386  386  386  385  385    0    0    0  386  386  385
           active       live_repair            active   387  387  386  386  385  385    0    0    0  386  386  385
           active       live_repair            active   387  387  386  386  385  385    0    0    0  386  386  385
           active       live_repair            active   387  387  386  386  385  385    0    0    0  386  386  385
           active       live_repair            active   387  387  386  386  385  385    0    0    0  386  386  385
       DS STATE 0        DS STATE 1        DS STATE 2  CON0 CON1 CON2 LRC0 LRC1 LRC2 LRA0 LRA1 LRA2 REP0 REP1 REP2
           active       live_repair            active   387  387  386  386  385  385    0    0    0  386  386  385
           active            active            active   387  387  386  386  386  385    0    0    0  386  386  385
           active            active            active   387  387  386  386  386  385    0    0    0  386  386  385
           active            active         replacing   387  387  386  386  386  385    0    0    0  386  386  386
           active            active         replacing   387  387  386  386  386  385    0    0    0  386  386  386
           active            active       live_repair   387  387  387  386  386  385    0    0    0  386  386  386
           active            active       live_repair   387  387  387  386  386  385    0    0    0  386  386  386
           active            active       live_repair   387  387  387  386  386  385    0    0    0  386  386  386
           active            active       live_repair   387  387  387  386  386  385    0    0    0  386  386  386
           active            active            active   387  387  387  386  386  386    0    0    0  386  386  386
           active            active            active   387  387  387  386  386  386    0    0    0  386  386  386
```

File: upstairs/src/lib.rs [hunk 1]
```diff
@@ -7287,11 +7398,74 @@ impl Upstairs {
...
+        }
+
+        // Now that we have verified our new target does not exist anywhere
+        // else, we can search for the old.
+        for (client_id, ds_target) in ds.ds_target.iter_mut().enumerate() {
+            if *ds_target == old {
+                info!(self.log, "{id} replacing old: {old} at {client_id}");
+                *ds_target = new;
+                // If downstairs_errors, clear them.
+                // clear region_metadata?  Yes, why not?  This makes the
```
Comment (jmpesp) [inline upstairs/src/lib.rs:7451]: I vote yes for proposition `clear region_metadata` :+1:

**Review by jmpesp — APPROVED** [review]
:shipit:

**Review by leftwo — COMMENTED** [review]
The recent changes should address all your comments.

File: upstairs/src/lib.rs [hunk 2]
```diff
@@ -7287,11 +7398,74 @@ impl Upstairs {
...
+        let active = self.active.lock().await;
+        let up_state = active.up_state;
+        let mut ds = self.downstairs.lock().await;
+
+        // ZZZ
+        // Block a replacement if any (other) downstairs are in:
+        // Replacing
+        // OnlineRepairReady
+        // OnlineRepair
+        // Faulted
```
[the remark this answers is not in the document]
Comment (leftwo) [reply upstairs/src/lib.rs]: Done

File: upstairs/src/lib.rs [hunk 3]
```diff
@@ -7411,61 +7411,107 @@ impl Upstairs {
...
+        // elsewhere, verified no other downstairs are in a bad state, we can
+        // move forward with the replacement.
+
+        info!(self.log, "{id} replacing old: {old} at {old_client_id}");
+        ds.ds_target[old_client_id] = new;
+
+        // If downstairs_errors, clear them.
+        // clear region_metadata?  Yes, why not?  This makes the
+        // code on the add side easier?  Is there anything that
+        // would prevent it?
```
Comment (jmpesp) [inline upstairs/src/lib.rs:7501]: this comment now needs updating

**Review by jmpesp — APPROVED** [review]
:rocket:

**Pull request merged**